Sara/mkt 388 integrating semgrep in ci (#683)
returntocorp/semgrep-docs: docs/semgrep-ci/overview.md

---
slug: overview
append_help_link: true
description: "Semgrep can run CI environments. It can either be used stand-alone or connected with Semgrep App for centralized rule and findings management."
hide_title: true
title: Getting started with Semgrep in CI 
---

import MoreHelp from "/src/components/MoreHelp"

# Getting started with Semgrep in continuous integration (CI)

Semgrep can be run within CI (continuous integration) environments. By integrating Semgrep into your CI environment, your development cycle benefits from the automated scanning of repositories at every push event, pull or merge request (PR or MR).

Semgrep is integrated into CI environments by creating a **job** (also known as an **action** for some CI providers) that is run by the CI provider.

A Semgrep CI job can **connect to Semgrep App**, a centralized location for managing findings, sending notifications, and configuring Semgrep rules for scanning. Alternatively, a Semgrep CI job can be set up as a **stand-alone** job, which means it does not connect to Semgrep App.

> This document aims to help users choose which method best meets their goals in integrating Semgrep in their CI environment.

Reviewer: ```suggestion
This document helps users choose which method best meets their goals in integrating Semgrep in their CI environment.
```

> After selecting a method, refer to the links in the [Next steps](#next-steps) section for documentation on how to set up or refine your Semgrep in CI job.

Reviewer: ```suggestion
After selecting a method, refer to the links in the [Next steps](#next-steps) section for documentation on how to set up or refine your CI job.
```

:::info

> When running in CI, Semgrep runs fully in your build environment. Your code is never sent anywhere.

Reviewer: ```suggestion
When running in CI, Semgrep runs fully in the CI build environment. Your code is never sent anywhere.
```

maybe?

> * Semgrep App collects [findings data](/docs/managing-findings/#semgrep-ci), which includes the line number of the code match, **not the code**. It is hashed using a one-way hashing function. Findings data is used to generate hyperlinks and support other Semgrep functions.

Reviewer: ```suggestion
* Semgrep App collects [findings data](/docs/managing-findings/#semgrep-ci), which includes the line number of the code match, but **not the code**. It is hashed using a one-way hashing function. Findings data is used to generate line-specific hyperlinks to your source code management system and support other Semgrep functions.
```

* CI scans that are not connected to Semgrep App do not send findings data to any servers.
:::

## Choosing a method

There are two methods to set up Semgrep in CI. 

> * To scale and automate your scans across many repositories, **use Semgrep App. **Semgrep App provides a centralized web application to manage findings and configure rulesets and notifications. This is the **recommended method** of integrating Semgrep in CI.

Reviewer: ```suggestion
* To scale and automate your scans across many repositories, **use Semgrep App**. Semgrep App provides a centralized web application to manage findings and configure rulesets and notifications. This is the **recommended method** of integrating Semgrep in CI.
```

This section describes a high-level view of steps to integrate Semgrep with Semgrep App.

![High level view of steps to integrate and refine Semgrep in your CI environment with Semgrep App](/img/semgrep-ci-overview-app.png "High level view of steps to integrate and refine Semgrep in your CI environment with Semgrep App")

_Figure 1._ High level view of steps to integrate and refine Semgrep in your CI environment with Semgrep App.

Integrating Semgrep in your CI environment through Semgrep App is the **recommended method** for CI providers such as GitHub Actions and GitLab CI/CD. 

The following steps outline the general procedure to automate Semgrep in many CI environments wish Semgrep App.

1. [Sign in](https://semgrep.dev/login) to Semgrep App.
2. Click on **Projects > Scan New Project**.
3. Follow instructions in the App. These steps vary based on your CI provider.
4. Semgrep App generates a `SEMGREP_APP_TOKEN` and creates a configuration file. 
5. Commit the configuration file and add `SEMGREP_APP_TOKEN` as a credential, token, or secret.
6. The scan automatically begins.
7. After the scan, Semgrep App reports findings (if any) into the Findings dashboard.

> 8. Optional: If you are not satisfied with your CI job's behavior, troubleshoot and edit the configuration file. Refer to [Sample CI configuration for Semgrep App](semgrep-ci/sample-ci-configs).

Reviewer: I got a 404 when following this link. I think the one you want is `/docs/semgrep-ci/sample-ci-configs/`.

9. Repeat from step 2 to integrate Semgrep into remaining repositories.

#### Default behaviors of scans set up through Semgrep App

> * Findings do not block a PR or MR. The presence of findings can be customized to block a PR or MR through the [Rule Board](docs/semgrep-app/rule-board).

Reviewer: I got a 404 when following this link. I think the one you want is `/docs/semgrep-app/rule-board/`.

> * The CI job does not fail when detecting findings, but can still fail due to other errors. See [Semgrep error codes](../cli-reference/#error-codes).

Reviewer: I got a 404 when following this link. I think the one you want is `/docs/cli-reference/#exit-codes`?
